pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"context"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+	"k8s.io/kubernetes/pkg/apis/discovery/validation"
+)
+
+// endpointSliceStrategy implements verification logic for Replication.
+type endpointSliceStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy is the default logic that applies when creating and updating Replication EndpointSlice objects.
+var Strategy = endpointSliceStrategy{legacyscheme.Scheme, names.SimpleNameGenerator}
+
+// NamespaceScoped returns true because all EndpointSlices need to be within a namespace.
+func (endpointSliceStrategy) NamespaceScoped() bool {
+	return true
+}
+
+// PrepareForCreate clears the status of an EndpointSlice before creation.
+func (endpointSliceStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+	endpointSlice := obj.(*discovery.EndpointSlice)
+	endpointSlice.Generation = 1
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (endpointSliceStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	newEPS := obj.(*discovery.EndpointSlice)
+	oldEPS := old.(*discovery.EndpointSlice)
+
+	// Increment generation if endpoints, ports, or address type have changed
```

Comment:
If you want to open-code the list of fields, add a comment that this needs to be updated if any fields are added to EndpointSlice

Reply:
That's why I suggested a white-out approach instead

Reply:
I did try a couple variations of that but kept on getting failing tests since the meta did not seem to be sticking, despite my efforts to reassign it after the whiteout. It didn't make any sense, will definitely revisit, just wanted to get tests back to passing for now.

Reply:
@liggitt @lavalamp Who has best context on use of `generation`?  Why don't we do it automatically?  Is there a safe best-practice other than open white-listing fields (likely to fail if new fields are added)?

Reply:
most types that use generation have a spec/status split, and do a deepequal comparison on the entire spec to decide whether to bump generation:

```
	if !apiequality.Semantic.DeepEqual(oldDaemonSet.Spec, newDaemonSet.Spec) {
		newDaemonSet.Generation = oldDaemonSet.Generation + 1
	}
```

```
	if !apiequality.Semantic.DeepEqual(oldRS.Spec, newRS.Spec) {
		newRS.Generation = oldRS.Generation + 1
	}
```


> Why don't we do it automatically?

things like annotations/labels/status generally shouldn't bump generation (especially status, otherwise status.observedGeneration would never be able to match generation)

Reply:
Ended up getting the white-out approach working, we'll just have to change/update that if we ever add something like a status field that we don't want to increment generation for.